chain/network/src/routing/store.rs

```diff
@@ -0,0 +1,206 @@
+use near_network_primitives::types::Edge;
+use near_primitives::network::PeerId;
+use smart_default::SmartDefault;
+use std::collections::{HashMap, HashSet};
+use std::io;
+use tracing::debug;
+
+/// TEST-ONLY
+#[derive(SmartDefault, Debug, Hash, PartialEq, Eq)]
+pub struct Component {
+    pub peers: Vec<PeerId>,
+    pub edges: Vec<Edge>,
+}
+
+/// Everytime a group of peers becomes unreachable at the same time; We store edges belonging to
+/// them in components. We remove all of those edges from memory, and save them to database,
+/// If any of them become reachable again, we re-add whole component.
+///
+/// To store components, we have following column in the DB.
+/// DBCol::LastComponentNonce -> stores component_nonce: u64, which is the lowest nonce that
+///                          hasn't been used yet. If new component gets created it will use
+///                          this nonce.
+/// DBCol::ComponentEdges     -> Mapping from `component_nonce` to list of edges
+/// DBCol::PeerComponent      -> Mapping from `peer_id` to last component nonce if there
+///                          exists one it belongs to.
+pub struct ComponentStore(schema::Store);
+impl ComponentStore {
+    pub fn new(store: near_store::Store) -> Self {
+        Self(schema::Store(store))
+    }
+
+    pub fn push_component(&self, peers: &HashSet<PeerId>, edges: &Vec<Edge>) -> io::Result<()> {
+        debug!(target: "network", "try_save_edges: We are going to remove {} peers", peers.len());
+        let component = self.0.get::<schema::LastComponentNonce>(&())?.unwrap_or(0) + 1;
+        let mut update = self.0.new_update();
+        update.set::<schema::LastComponentNonce>(&(), &component);
+        update.set::<schema::ComponentEdges>(&component, &edges);
+        for peer_id in peers {
+            update.set::<schema::PeerComponent>(peer_id, &component);
+        }
+        update.commit()
+    }
```

Comment:
Ok, so that's a design pattern that I used to apply in my previous projects:
1. one abstraction layer to define the schema and make the DB access strongly typed - we don't want any custom code here, it should be as declarative as possible and as concise as possible (think of it as an autogenerated layer from a DB schema). We accept that calls to that layer are not perfectly pretty.
2. custom (potentially multiple) views for accessing the DB implemented on top of layer described above. They can be used to make the access prettier, but first of all they allow to enforce high level invariants, like the relations between columns, etc.

Reply:
Ah, I see, "relationships between columns" is exactly what we are doing here. Thanks, that's a useful mental model!